crates/bevy_ecs/src/schedule/schedule.rs

```diff
@@ -844,19 +876,107 @@ impl ScheduleGraph {
 
     /// Initializes any newly-added systems and conditions by calling [`System::initialize`]
     pub fn initialize(&mut self, world: &mut World) {
-        for (id, i) in self.uninit.drain(..) {
-            match id {
-                NodeId::System(index) => {
-                    self.systems[index].get_mut().unwrap().initialize(world);
+        for uninit in self.uninit.drain(..) {
+            match uninit {
+                PendingInit::System(index) => {
+                    let id = NodeId::System(index);
+                    let system = self.systems[index].get_mut().unwrap();
+                    system.initialize(world);
+
+                    let system_access = system.component_access();
+                    for &set in system_access
+                        .reads()
+                        .filter_map(|id| self.read_system_sets.get(id))
+                        .flatten()
+                    {
+                        self.hierarchy.graph.add_edge(set, id, ());
+                    }
+                    for &set in system_access
+                        .writes()
+                        .filter_map(|id| self.write_system_sets.get(id))
+                        .flatten()
+                    {
+                        self.hierarchy.graph.add_edge(set, id, ());
+                    }
+
                     for condition in &mut self.system_conditions[index] {
                         condition.initialize(world);
                     }
                 }
-                NodeId::Set(index) => {
+                PendingInit::SetCondition {
+                    set_index: index,
+                    condition_index: i,
+                } => {
                     for condition in self.system_set_conditions[index].iter_mut().skip(i) {
                         condition.initialize(world);
                     }
                 }
+                PendingInit::AccessSet(index) => {
+                    let id = NodeId::Set(index);
+                    let set = self.system_sets[index].inner.as_ref();
+                    if let Some(component) = set.reads_component() {
+                        let component = world
+                            .components
+                            .init_component_with_descriptor(&mut world.storages, component);
+                        self.read_system_sets
+                            .get_or_insert_with(component, Vec::new)
+                            .push(id);
+
+                        // If true, that means there are old systems that need to be initialized.
+                        let mut init_old = false;
```

Comment:
(i'm new to bevy and i'm going through this PR to understand better the ECS)

Why can there be old systems that need to be initialized?
Is it be cause the AccessSet initialization runs before some system initialization?
Would it make sense to always make sure that AccessSet initialization runs at the very end of the initialization?

Reply:
It's to handle a situation like this

```rust
schedule.add_systems(my_system);
schedule.run(&must world);
schedule.add_systems(other_system.after(Writes::<T>));
```